### ISSUE TITLE
Performance issue: loading files on every find call?

### ISSUE THREAD

**Issue by magwo** [evansiroky/node-geo-tz#46]
Am I correct if I say that the code loads a feature file on most find calls?

https://github.com/evansiroky/node-geo-tz/blob/master/lib/find.js#L88

Also, it does this using readFileSync which is a slight problem but not really my concern, and it's nice that the API is sync.

Anyway... my use case involves quickly looking up the timezone of thousands of positions in one go. Looking up 10k positions appears to take about 6 seconds on a computer here, which is ok but not great for me. I was unable to find anything else in the code that is apparently unperformant, so I'm fairly sure it's the file loading that is the bottleneck. (Could/should profile though).

Assuming the file loading is the slowdown, to make this faster, I have a suggestion: To the find() function optionally provide a "feature provider". Two such feature providers would be added to the library: OnDemandFeatureProvider (which would load .buf files on demand) and a PreloadedFeatureProvider, which would exhaustively preload and decode all the .buf files upon instantiation, and then be able to provide pure geoJson objects directly which would probably be several magnitudes faster than the file loading. The downside being much higher memory usage. However, given this provider pattern, a library user can decide if they want to trade memory for speed or not.

How much memory do you think we're talking about here, to keep all the preloaded features in memory as geoJson objects?

Let me know if such a PR would be interesting, or if you would like to use a different architecture to provide optional pre-loading of data.

**Comment by evansiroky** [comment]
Processing certain does go faster when everything is in memory.  I think for most modern computing machines that it would be possible to have everything in memory.  In a previous version of this library I was loading a GeoJSON file of the world into memory it certainly did use a lot of memory, maybe 500mb - don't remember the exact amount.  

I'd welcome a PR for this and believe that it would be possible to preserve the synchronous nature of the API.  In order to keep the memory under control like the existing API does, I think that it would be best to have an all-or-nothing approach of choosing how much data to keep in memory.  If one wants to use the Preloaded data, perhaps there could be a method that loads everything into memory at the beginning.  And to make that more performant, it might be best to modify the updating part of this library to save all of the indexed data that would get preloaded into a single file.

Something to keep in mind is that this project uses JavaScript which is single-threaded and generally not super-performant.  Ultimately if you are looking for much faster performance you might want to use GIS software and use the raw timezone data directly.  You can find the timezone shapefiles and GeoJSON here: https://github.com/evansiroky/timezone-boundary-builder/releases

**Comment by magwo** [comment]
Ok cool. I'm working on a PR. Will try to add some kind of performance test in addition to the correctness tests for proof of performance improvement.

**Comment by magwo** [comment]
By the way I have three failing tests on master.. should I just ignore them?

**Comment by evansiroky** [comment]
The update test has been flaky due to problems with the nock library.  All the other ones ought to work fine.

**Comment by evansiroky** [comment]
If any other test than the "data update > tz geojson should get updated after fetching valid shapefile" fail, please submit a bug report.

**Comment by magwo** [comment]
By the way, an approximating recursive object size calculation says the full data set is about 140 MB in-memory.

**Comment by magwo** [comment]
Will see if the tests still fail when I make the PR.

**Comment by magwo** [comment]
https://github.com/evansiroky/node-geo-tz/pull/47

**Comment by magwo** [comment]
Some example runs:
```find tz of 2000 random european positions with unspecified: 5499ms
find tz of 2000 random european positions with preloaded: 168ms

find tz of 2000 random european positions with unspecified: 4443ms
find tz of 2000 random european positions with preloaded: 173ms

find tz of 2000 random european positions with unspecified: 4584ms
find tz of 2000 random european positions with preloaded: 146ms
```

**Comment by magwo** [comment]
Actually, using the preloaded one bumps memory usage by about 900 MB.. which is a problem. :)

Uhm..

**Comment by magwo** [comment]
What do you think about Google's polyline encoding? It's extremely compact (~5% size I believe compared to geoJson) and fast to decode and encode. https://developers.google.com/maps/documentation/utilities/polylinealgorithm

However, maybe this PR is the wrong place to introduce polyline encoding? Maybe that should be done by the boundary builder? I will do an experiment to see the performance of encoding the feature data in polyline and decoding on demand when boundary checking the features.

**Comment by evansiroky** [comment]
The polyline encoding idea might save a little bit of memory.  The 900 MB of memory usage seems believable and well within the abilities for most machines to handle.  I think the extra memory usage is a reasonable tradeoff to yield faster lookup times which is the ultimate goal.

**Comment by evansiroky** [comment]
Fixed with #47
